File: csv_export.py
"""
CSV Export module - Export data to CSV format
"""
import csv
import json


class CSVExporter:
    """Export OSM elevation data to CSV"""
    
    @staticmethod
    def get_element_info(element):
        """Extract relevant information from OSM element"""
        element_type = element.get("type", "")
        element_id = element.get("id", "")
        
        # Get coordinates
        if element_type == "node":
            lat = element.get("lat", "")
            lon = element.get("lon", "")
        elif element_type == "way" and "center" in element:
            lat = element["center"].get("lat", "")
            lon = element["center"].get("lon", "")
        else:
            lat = ""
            lon = ""
        
        # Get tags
        tags = element.get("tags", {})
        name = tags.get("name", tags.get("ref", ""))
        ele = tags.get("ele", "")
        ele_source = tags.get("ele:source", "")
        tourism = tags.get("tourism", "")
        railway = tags.get("railway", "")
        
        return {
            "type": element_type,
            "id": element_id,
            "name": name,
            "lat": lat,
            "lon": lon,
            "elevation": ele,
            "elevation_source": ele_source,
            "tourism": tourism,
            "railway": railway,
            "osm_link": f"https://www.openstreetmap.org/{element_type}/{element_id}"
        }
    
    def export_to_csv(self, data, output_file):
        """
        Export data to CSV file
        
        Args:
            data: Dict with categories of OSM elements
            output_file: Output CSV filename
        """
        rows = []
        
        # Process all categories
        for category, elements in data.items():
            # Handle both list of elements and dict with valid_elements
            if isinstance(elements, dict) and "valid_elements" in elements:
                elements = elements["valid_elements"]
            
            if not isinstance(elements, list):
                continue
            
            for element in elements:
                info = self.get_element_info(element)
                info["category"] = category
                rows.append(info)
        
        # Write to CSV
        if rows:
            fieldnames = [
                "category", "type", "id", "name", "lat", "lon", 
                "elevation", "elevation_source", "tourism", "railway", "osm_link"
            ]
            
            with open(output_file, "w", newline="", encoding="utf-8") as csvfile:
                writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
                writer.writeheader()
                writer.writerows(rows)
            
            print(f"Exported {len(rows)} elements to {output_file}")
            return len(rows)
        else:
            print("No data to export")
            return 0


if __name__ == "__main__":
    # Load validated data
    with open("osm_data_validated.json", "r", encoding="utf-8") as f:
        data = json.load(f)
    
    # Export to CSV
    exporter = CSVExporter()
    exporter.export_to_csv(data, "elevation_data.csv")
    print("CSV export complete!")

File: src/csv_export.go
package main

import (
	"encoding/csv"
	"fmt"
	"os"
	"strconv"
)

type CSVExporter struct{}

type ElementInfo struct {
	Category        string
	Type            string
	ID              string
	Name            string
	Lat             string
	Lon             string
	Elevation       string
	ElevationSource string
	Tourism         string
	Railway         string
	OSMLink         string
}

func NewCSVExporter() *CSVExporter {
	return &CSVExporter{}
}

func (e *CSVExporter) getElementInfo(element OSMElement, category string) ElementInfo {
	info := ElementInfo{
		Category: category,
		Type:     element.Type,
		ID:       strconv.FormatInt(element.ID, 10),
	}

	// Get coordinates
	if element.Type == "node" {
		info.Lat = fmt.Sprintf("%.6f", element.Lat)
		info.Lon = fmt.Sprintf("%.6f", element.Lon)
	} else if element.Type == "way" && element.Center != nil {
		info.Lat = fmt.Sprintf("%.6f", element.Center.Lat)
		info.Lon = fmt.Sprintf("%.6f", element.Center.Lon)
	}

	// Get tags
	if element.Tags != nil {
		if name, ok := element.Tags["name"]; ok {
			info.Name = name
		} else if ref, ok := element.Tags["ref"]; ok {
			info.Name = ref
		}

		info.Elevation = element.Tags["ele"]
		info.ElevationSource = element.Tags["ele:source"]
		info.Tourism = element.Tags["tourism"]
		info.Railway = element.Tags["railway"]
	}

	// OSM link
	info.OSMLink = fmt.Sprintf("https://www.openstreetmap.org/%s/%d", element.Type, element.ID)

	return info
}

func (e *CSVExporter) ExportToCSV(data ValidatedData, outputFile string) (int, error) {
	var rows []ElementInfo

	// Process all categories
	categories := map[string][]OSMElement{
		"train_stations":       data.TrainStations.ValidElements,
		"alpine_huts":          data.AlpineHuts.ValidElements,
		"other_accommodations": data.OtherAccommodations.ValidElements,
	}

	for category, elements := range categories {
		for _, element := range elements {
			info := e.getElementInfo(element, category)
			rows = append(rows, info)
		}
	}

	if len(rows) == 0 {
		fmt.Println("No data to export")
		return 0, nil
	}

	// Create CSV file
	file, err := os.Create(outputFile)
	if err != nil {
		return 0, fmt.Errorf("failed to create CSV file: %v", err)
	}
	defer file.Close()

	writer := csv.NewWriter(file)
	defer writer.Flush()

	// Write header
	header := []string{
		"category", "type", "id", "name", "lat", "lon",
		"elevation", "elevation_source", "tourism", "railway", "osm_link",
	}
	if err := writer.Write(header); err != nil {
		return 0, fmt.Errorf("failed to write header: %v", err)
	}

	// Write rows
	for _, row := range rows {
		record := []string{
			row.Category,
			row.Type,
			row.ID,
			row.Name,
			row.Lat,
			row.Lon,
			row.Elevation,
			row.ElevationSource,
			row.Tourism,
			row.Railway,
			row.OSMLink,
		}
		if err := writer.Write(record); err != nil {
			return 0, fmt.Errorf("failed to write row: %v", err)
		}
	}

	fmt.Printf("Exported %d elements to %s\n", len(rows), outputFile)
	return len(rows), nil
}

func runExportCSV() error {
	fmt.Println("\n" + string(repeat('=', 60)))
	fmt.Println("STEP 5: EXPORT - Creating CSV output")
	fmt.Println(string(repeat('=', 60)))

	// Load validated data
	var data ValidatedData
	if err := loadJSON("output/osm_data_validated.json", &data); err != nil {
		return fmt.Errorf("output/osm_data_validated.json not found. Run --validate first: %v", err)
	}

	// Export to CSV
	exporter := NewCSVExporter()
	count, err := exporter.ExportToCSV(data, "output/elevation_data.csv")
	if err != nil {
		return err
	}

	fmt.Printf("\n✓ Exported %d elements to output/elevation_data.csv\n\n", count)

	return nil
}
